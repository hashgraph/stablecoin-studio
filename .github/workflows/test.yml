name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test