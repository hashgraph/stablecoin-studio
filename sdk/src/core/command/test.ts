/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
