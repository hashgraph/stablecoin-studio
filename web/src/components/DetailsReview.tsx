import type { FlexProps as ChakraFlexProps, TextProps as ChakraTextProps } from '@chakra-ui/react';
import {
	Box,
	Button,
	Divider,
	Flex,
	HStack,
	Link,
	Stack,
	Text,
	useBoolean,
	useDisclosure,
} from '@chakra-ui/react';
import type {
	AccountViewModel,
	RequestPublicKey,
	UpdateRequest,
} from '@hashgraph/stablecoin-npm-sdk';
import { Account } from '@hashgraph/stablecoin-npm-sdk';
import { useEffect, useState } from 'react';
import type { FieldValues, UseFormGetValues } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { SDKService } from '../services/SDKService';
import ModalsHandler from './ModalsHandler';
import type { ModalsHandlerActionsProps } from './ModalsHandler';
import { SELECTED_WALLET_ACCOUNT_INFO, SELECTED_WALLET_COIN } from '../store/slices/walletSlice';
import AwaitingWalletSignature from './AwaitingWalletSignature';
import DatePickerController from './Form/DatePickerController';
import InputController from './Form/InputController';
import InputNumberController from './Form/InputNumberController';
import { RouterManager } from '../Router/RouterManager';
import Icon from './Icon';
import { KeySelector } from './KeySelector';
import TooltipCopy from './TooltipCopy';
import MaskData from 'maskdata';

export interface Detail {
	label: string;
	labelInBold?: boolean;
	value: any; // TODO: string | number
	valueInBold?: boolean;
	copyButton?: boolean;
	hashScanURL?: string;
}

export interface DetailsReviewProps {
	details: Detail[];
	divider?: boolean;
	title?: string | null;
	titleProps?: ChakraTextProps;
	contentProps?: ChakraFlexProps;
	isLoading?: boolean;
	editable?: boolean;
	getStableCoinDetails?: () => Promise<void>;
}

const allowedKeys = ['KYC key', 'Freeze key', 'Wipe key', 'Pause key', 'Fee schedule key'];

let fieldsCanEdit = [
	{
		id: 'Name',
		type: 'text',
	},
	{
		id: 'Symbol',
		type: 'text',
	},
	{
		id: 'Expiration time',
		type: 'date',
	},
	{
		id: 'Autorenew period',
		type: 'number',
	},
	{
		id: 'Metadata',
		type: 'text',
	},
];

const commonTextProps: ChakraTextProps = {
	fontSize: '14px',
	fontWeight: 500,
	lineHeight: '17px',
	color: 'brand.gray',
	wordBreak: 'break-all',
};

const textInBoldProps: ChakraTextProps = {
	...commonTextProps,
	fontWeight: 700,
	color: 'brand.black',
};

const daysToSeconds = (days: number): string => {
	return (days * 24 * 60 * 60).toString();
};

const formatKey = (
	keySelection: string,
	keyName: string,
	accountInfo: AccountViewModel,
	getValues: UseFormGetValues<FieldValues>,
): RequestPublicKey | undefined => {
	const values = getValues();

	if (keySelection === 'Current user key') {
		return accountInfo.publicKey;
	}

	if (keySelection === 'Other key') {
		const param = Object.keys(values).find((key) => key.includes(keyName + 'Other'));

		return {
			key: param ? values[param] : '',
		};
	}

	if (keySelection === 'None') return undefined;

	return Account.NullPublicKey;
};

const keyToString = (
	keySelection: any,
	keyName: string,
	accountInfo: AccountViewModel,
	getValues: UseFormGetValues<FieldValues>,
): string => {
	if (keySelection) {
		if (keySelection.label === 'Current user key' || keySelection.label === 'The smart contract')
			return keySelection.label;
		return maskPublicKeys(formatKey(keySelection.label, keyName, accountInfo, getValues)!.key);
	} else {
		return 'None';
	}
};

const maskPublicKeys = (key: string): string => {
	const maskJSONOptions = {
		maskWith: '.',
		unmaskedStartCharacters: 4,
		unmaskedEndCharacters: 4,
	};
	return MaskData.maskPassword(key, maskJSONOptions);
};

const DetailsReview = ({
	details,
	divider = true,
	title,
	titleProps,
	contentProps,
	isLoading,
	editable = false,
	getStableCoinDetails,
}: DetailsReviewProps) => {
	const {
		isOpen: isOpenModalAction,
		onOpen: onOpenModalAction,
		onClose: onCloseModalAction,
	} = useDisclosure();

	const { control, setValue, reset, getValues } = useForm({
		mode: 'onChange',
	});

	const [errorOperation, setErrorOperation] = useState();
	const [errorTransactionUrl, setErrorTransactionUrl] = useState();

	const navigate = useNavigate();
	const handleCloseModal = () => {
		RouterManager.goBack(navigate);
	};

	const [editMode, setEditMode] = useBoolean(false);
	const accountInfo = useSelector(SELECTED_WALLET_ACCOUNT_INFO);
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);

	useEffect(() => {
		fieldsCanEdit = fieldsCanEdit.filter((field) => !field.id.includes('key'));
		editable &&
			details.forEach((item) => {
				if (allowedKeys.includes(item.label) && item.value !== 'NONE')
					fieldsCanEdit.push({
						id: item.label,
						type: 'key',
					});
			});
	}, [details]);

	const handleEditMode = () => {
		if (!editMode) {
			setValue('name', details.find((item) => item.label === 'Name')?.value);
			setValue('symbol', details.find((item) => item.label === 'Symbol')?.value);
			setValue(
				'autorenew period',
				details.find((item) => item.label === 'Autorenew period')?.value,
			);
			setValue('expiration time', details.find((item) => item.label === 'Expiration time')?.value);
			setValue('metadata', details.find((item) => item.label === 'Metadata')?.value);

			fieldsCanEdit.find((fields) => fields.id === 'Freeze key') &&
				setValue(
					'freeze key',
					formatDetailToKey(details.find((item) => item.label === 'Freeze key')!),
				);

			fieldsCanEdit.find((fields) => fields.id === 'Pause key') &&
				setValue(
					'pause key',
					formatDetailToKey(details.find((item) => item.label === 'Pause key')!),
				);
			fieldsCanEdit.find((fields) => fields.id === 'Wipe key') &&
				setValue('wipe key', formatDetailToKey(details.find((item) => item.label === 'Wipe key')!));
			fieldsCanEdit.find((fields) => fields.id === 'KYC key') &&
				setValue('kyc key', formatDetailToKey(details.find((item) => item.label === 'KYC key')!));
			fieldsCanEdit.find((fields) => fields.id === 'Fee schedule key') &&
				setValue(
					'fee schedule key',
					formatDetailToKey(details.find((item) => item.label === 'Fee schedule key')!),
				);
		} else {
			reset();
		}

		setEditMode.toggle();
	};

	const formatDetailToKey = ({
		value,
		label,
		hashScanURL,
	}: {
		value: string;
		label: string;
		hashScanURL?: string;
	}) => {
		if (value.includes('SMART')) {
			return { value: 1, label: 'The smart contract' };
		}

		if (value.substring(0, 6) === accountInfo?.publicKey?.key.substring(0, 6)) {
			return { value: 2, label: 'Current user key' };
		}

		setValue(`${label.toLowerCase()}Other`, hashScanURL?.split('account/')[1]);
		return { value: 3, label: 'Other key' };
	};

	if (isLoading && editMode) return <AwaitingWalletSignature />;

	const { t } = useTranslation(['global', 'stableCoinDetails', 'updateToken']);
	const currentExpirationTime: Detail | undefined = details.find(
		(obj) => obj.label === 'Expiration time',
	);
	const maximumExpirationTime: Date = currentExpirationTime
		? new Date(currentExpirationTime.value)
		: new Date();
	maximumExpirationTime.setFullYear(maximumExpirationTime.getFullYear() + 5);

	const handleUpdateToken: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
	}) => {
		try {
			onLoading();
			if (
				!selectedStableCoin?.proxyAddress ||
				!selectedStableCoin?.proxyAdminAddress ||
				!selectedStableCoin?.tokenId
			) {
				onError();
				return;
			}
			// @ts-ignore
			const request: UpdateRequest = {
				tokenId: details.find((detail) => detail.label === 'Token id')?.value as string,
				name: getValues().name,
				symbol: getValues().symbol,
				metadata: getValues().metadata,
				autoRenewPeriod: daysToSeconds(getValues()['autorenew period']),
				expirationTimestamp: Date.parse(getValues()['expiration time']) * 1000000 + '',
			};
			if (getValues()['freeze key']) {
				request.freezeKey = formatKey(
					getValues()['freeze key'].label,
					'freeze key',
					accountInfo,
					getValues,
				);
			}
			if (getValues()['wipe key']) {
				request.wipeKey = formatKey(
					getValues()['wipe key'].label,
					'wipe key',
					accountInfo,
					getValues,
				);
			}
			if (getValues()['pause key']) {
				request.pauseKey = formatKey(
					getValues()['pause key'].label,
					'pause key',
					accountInfo,
					getValues,
				);
			}
			if (getValues()['kyc key']) {
				request.kycKey = formatKey(getValues()['kyc key'].label, 'kyc key', accountInfo, getValues);
			}
			if (getValues()['fee schedule key']) {
				request.feeScheduleKey = formatKey(
					getValues()['fee schedule key'].label,
					'fee schedule key',
					accountInfo,
					getValues,
				);
			}
			request.validate = () => [];
			await SDKService.updateStableCoin(request);
			setTimeout(() => {
				setEditMode.off();
				getStableCoinDetails && getStableCoinDetails();
			}, 3000);
			onSuccess();
		} catch (error: any) {
			console.log(JSON.stringify(error));
			setErrorTransactionUrl(error.transactionUrl);
			setErrorOperation(error.message);
			onError();
		}
	};

	return (
		<Box textAlign='left' data-testid='button-box'>
			<Flex justify='space-between' mb={6} alignItems='center'>
				{title && (
					<Text
						data-testid='details-review-title'
						fontSize='16px'
						fontWeight={600}
						color='brand.gray'
						{...titleProps}
					>
						{title}
					</Text>
				)}
				{editable && (
					<>
						{editMode ? (
							<Icon
								fontSize={16}
								name='X'
								onClick={handleEditMode}
								_hover={{ cursor: 'pointer' }}
							/>
						) : (
							<Icon
								fontSize={16}
								name='Pen'
								onClick={handleEditMode}
								_hover={{ cursor: 'pointer' }}
							/>
						)}
					</>
				)}
			</Flex>
			<Box>
				{details.map((detail: Detail, index: number, details: Detail[]) => (
					<Box key={`details-review-detail-${index}`}>
						<Flex
							data-testid={`details-review-detail-${index}`}
							justifyContent='space-between'
							{...contentProps}
						>
							<Text
								whiteSpace={'nowrap'}
								{...(detail.labelInBold ? textInBoldProps : commonTextProps)}
							>
								{detail.label}
							</Text>
							{editMode && fieldsCanEdit.map((field) => field.id).includes(detail.label) ? (
								<>
									{fieldsCanEdit.find((field) => field.id === detail.label)?.type === 'text' && (
										<InputController
											control={control}
											name={detail.label.toLowerCase()}
											placeholder={detail.label}
											containerStyle={{
												w: '300px',
											}}
										/>
									)}
									{fieldsCanEdit.find((field) => field.id === detail.label)?.type === 'number' && (
										<InputNumberController
											rules={{
												validate: {
													validation: (value: number) => {
														const res = t('updateToken:updatingValidation.autoRenewPeriod');
														return detail.label === t('stableCoinDetails:autoRenewPeriod') &&
															(value > 92 || value < 30)
															? res
															: true;
													},
												},
											}}
											control={control}
											name={detail.label.toLowerCase()}
											placeholder={detail.label}
											containerStyle={{
												w: '300px',
											}}
										/>
									)}
									{fieldsCanEdit.find((field) => field.id === detail.label)?.type === 'date' && (
										<DatePickerController
											control={control}
											name={detail.label.toLowerCase()}
											placeholder={detail.label}
											containerStyle={{
												w: '300px',
											}}
											minimumDate={new Date(detail.value)}
											maximumDate={maximumExpirationTime}
											customHeader={false}
											showMonthDropdown
											showYearDropdown
										/>
									)}
									{fieldsCanEdit.find((field) => field.id === detail.label)?.type === 'key' && (
										<Stack w='300px'>
											<KeySelector
												control={control}
												name={detail.label.toLowerCase()}
												label=''
												labelPlaceholder={detail.label}
											/>
										</Stack>
									)}
								</>
							) : (
								<>
									{typeof detail.value === 'string' && detail.value.includes('GMT') ? (
										<Text {...commonTextProps}>{new Date(detail.value).toDateString()}</Text>
									) : typeof detail.value === 'string' || typeof detail.value === 'number' ? (
										<HStack {...(detail.valueInBold ? textInBoldProps : commonTextProps)}>
											<Text>{detail.value.toString()}</Text>
											{detail.copyButton && (
												<TooltipCopy valueToCopy={detail.value.toString() ?? ''}>
													<Icon name='Copy' />
												</TooltipCopy>
											)}
											{detail.hashScanURL && (
												<Link isExternal={true} href={`${detail.hashScanURL}`}>
													<Icon name='ArrowSquareOut' />
												</Link>
											)}
										</HStack>
									) : detail.value && 'toString' in detail.value && 'value' in detail.value ? (
										detail.value.toString()
									) : (
										detail.value
									)}
								</>
							)}
						</Flex>
						{divider && details.length !== index + 1 && (
							<Divider title='divider' mt='13px' mb='13px' />
						)}
					</Box>
				))}
				{editMode && (
					<HStack mt={10} justify={'flex-end'}>
						<Button onClick={handleEditMode} variant='secondary'>
							Cancel
						</Button>
						{/* @ts-ignore TODO: add T generic in controllers */}
						<Button onClick={onOpenModalAction}>Save</Button>
					</HStack>
				)}
			</Box>

			<ModalsHandler
				errorNotificationTitle={t('operations:modalErrorTitle')}
				errorNotificationDescription={errorOperation}
				errorTransactionUrl={errorTransactionUrl}
				modalActionProps={{
					isOpen: isOpenModalAction,
					onClose: onCloseModalAction,
					title: t('updateToken:modalAction.subtitle'),
					confirmButtonLabel: t('updateToken:modalAction.accept'),
					onConfirm: handleUpdateToken,
				}}
				ModalActionChildren={
					<DetailsReview
						title={t('updateToken:modalAction.subtitle')}
						details={[
							{
								label: t('stableCoinDetails:name'),
								value: getValues().name,
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:symbol'),
								value: getValues().symbol,
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:metadata'),
								value: getValues().metadata,
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:autoRenewPeriod'),
								value: getValues()['autorenew period'],
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:expirationTime'),
								value: new Date(Date.parse(getValues()['expiration time'])).toString(),
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:kycKey'),
								value: keyToString(getValues()['kyc key'], 'kyc key', accountInfo, getValues),
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:freezeKey'),
								value: keyToString(getValues()['freeze key'], 'freeze key', accountInfo, getValues),
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:wipeKey'),
								value: keyToString(getValues()['wipe key'], 'wipe key', accountInfo, getValues),
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:pauseKey'),
								value: keyToString(getValues()['pause key'], 'pause key', accountInfo, getValues),
								valueInBold: true,
							},
							{
								label: t('stableCoinDetails:feeScheduleKey'),
								value: keyToString(
									getValues()['fee schedule key'],
									'fee schedule key',
									accountInfo,
									getValues,
								),
								valueInBold: true,
							},
						]}
					/>
				}
				successNotificationTitle={t('operations:modalSuccessTitle')}
				successNotificationDescription={t('operations:modalSuccessDesc')}
				handleOnCloseModalError={handleCloseModal}
				handleOnCloseModalSuccess={handleCloseModal}
			/>
		</Box>
	);
};

export default DetailsReview;
