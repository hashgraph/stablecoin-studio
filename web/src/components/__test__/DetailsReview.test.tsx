import DetailsReview from '../DetailsReview';
import { render } from '../../test/index';

const details = [
	{
		label: 'Account to revoke rol',
		value: '0.0.734953',
	},
	{
		label: 'Role to revoke',
		value: 'Rescuer',
		valueInBold: true,
	},
];

const defaultProps = {
	details,
	title: 'Datails Review',
};

describe(`<${DetailsReview.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<DetailsReview {...defaultProps} />);

		expect(component).toMatchSnapshot();
	});

	test('should render title if exists', () => {
		const component = render(<DetailsReview {...defaultProps} />);

		const title = component.getByTestId('details-review-title');
		expect(title).toBeInTheDocument();
		expect(title).toHaveTextContent(defaultProps.title);
	});

	test('should not render title if it is not in props', () => {
		const component = render(<DetailsReview details={details} />);

		const title = component.queryByTestId('details-review-title');
		expect(title).not.toBeInTheDocument();
	});

	test('should render all details', () => {
		const component = render(<DetailsReview {...defaultProps} />);
		const { details } = defaultProps;

		details.forEach((item, index) => {
			const detail = component.getByTestId(`details-review-detail-${index}`);
			expect(detail).toHaveTextContent(item.label);
			expect(detail).toHaveTextContent(item.value);
		});
	});

	test('should has a divider between details by default', () => {
		const detail = {
			label: 'testing',
			value: 'testing',
		};
		const details = Array.from({ length: 15 }, () => detail);
		const props = {
			details,
			title: 'Long details list for testing',
		};
		const component = render(<DetailsReview {...props} />);

		const dividers = component.getAllByTitle('divider');
		expect(dividers).toHaveLength(details.length - 1);
	});

	test('should not has dividers is divider prop is false', () => {
		const component = render(<DetailsReview divider={false} {...defaultProps} />);

		const dividers = component.queryAllByTitle('divider');
		expect(dividers).toHaveLength(0);
	});
});
