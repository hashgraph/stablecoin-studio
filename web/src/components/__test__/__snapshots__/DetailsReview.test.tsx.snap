// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsReview /> should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <div
        class="css-13brihr"
        data-testid="button-box"
      >
        <div
          class="css-1i64jsd"
        >
          <p
            class="chakra-text css-psqrpk"
            data-testid="details-review-title"
          >
            Datails Review
          </p>
        </div>
        <div
          class="css-0"
        >
          <div
            class="css-0"
          >
            <div
              class="css-gg4vpm"
              data-testid="details-review-detail-0"
            >
              <p
                class="chakra-text css-9s6br4"
              >
                Account to revoke rol
              </p>
              <div
                class="chakra-stack css-1c8ysgd"
              >
                <p
                  class="chakra-text css-mgupmj"
                >
                  0.0.734953
                </p>
              </div>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-zaaihy"
              title="divider"
            />
          </div>
          <div
            class="css-0"
          >
            <div
              class="css-gg4vpm"
              data-testid="details-review-detail-1"
            >
              <p
                class="chakra-text css-9s6br4"
              >
                Role to revoke
              </p>
              <div
                class="chakra-stack css-9qi64u"
              >
                <p
                  class="chakra-text css-mgupmj"
                >
                  Rescuer
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="css-13brihr"
      data-testid="button-box"
    >
      <div
        class="css-1i64jsd"
      >
        <p
          class="chakra-text css-psqrpk"
          data-testid="details-review-title"
        >
          Datails Review
        </p>
      </div>
      <div
        class="css-0"
      >
        <div
          class="css-0"
        >
          <div
            class="css-gg4vpm"
            data-testid="details-review-detail-0"
          >
            <p
              class="chakra-text css-9s6br4"
            >
              Account to revoke rol
            </p>
            <div
              class="chakra-stack css-1c8ysgd"
            >
              <p
                class="chakra-text css-mgupmj"
              >
                0.0.734953
              </p>
            </div>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-zaaihy"
            title="divider"
          />
        </div>
        <div
          class="css-0"
        >
          <div
            class="css-gg4vpm"
            data-testid="details-review-detail-1"
          >
            <p
              class="chakra-text css-9s6br4"
            >
              Role to revoke
            </p>
            <div
              class="chakra-stack css-9qi64u"
            >
              <p
                class="chakra-text css-mgupmj"
              >
                Rescuer
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
